rpc/client.go

```diff
@@ -26,212 +26,606 @@ type Transport interface {
+	client.Transport = config.Transport
+
+	return client, nil
+}
 
-	go client.listen(client.transport)
+type Client struct {
+	Logger    log.Logger
+	Transport Transport
+
+	eventHandlers   map[Event]Handler
+	commandChannels map[string]chan responseMessage
+	clientID        snowflake.ID
+
+	readyChan    chan struct{}
+	User         discord.User
+	ServerConfig ServerConfig
+	V            int
+}
+
+func (c *Client) Open() error {
+	go c.listen(c.Transport)
 
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
 	select {
 	case <-ctx.Done():
-		return nil, ctx.Err()
-	case <-client.readyChan:
+		return ctx.Err()
+	case <-c.readyChan:
 	}
+	return nil
+}
 
-	return client, nil
+func (c *Client) sendBytes(r io.Reader) error {
+	writer, err := c.Transport.NextWriter()
+	if err != nil {
+		return err
+	}
+	defer writer.Close()
+
+	buff := new(bytes.Buffer)
+	newWriter := io.MultiWriter(writer, buff)
+
+	_, err = io.Copy(newWriter, r)
+	if err != nil {
+		return err
+	}
+
+	data, _ := io.ReadAll(buff)
+	c.Logger.Tracef("Sending message: data: %s", string(data))
+
+	return err
 }
 
-type clientImpl struct {
-	logger    log.Logger
-	transport Transport
+func (c *Client) Authorize(scopes []discord.OAuth2Scope, rpcToken string, username string) (string, error) {
+	args := CmdArgsAuthorize{
+		ClientID: c.clientID,
+		Scopes:   scopes,
+		RPCToken: rpcToken,
+		Username: username,
+	}
 
-	eventHandlers   map[Event]Handler
-	commandHandlers map[string]internalHandler
+	if res, err := c.send(Message{
+		Cmd:  CmdAuthorize,
+		Args: args,
+	}); err != nil {
+		return "", err
+	} else {
+		return res.(CmdRsAuthorize).Code, nil
+	}
+}
 
-	readyChan    chan struct{}
-	user         discord.User
-	serverConfig ServerConfig
-	v            int
+func (c *Client) Authenticate(accessToken string) (CmdRsAuthenticate, error) {
+	if res, err := c.send(Message{
+		Cmd:  CmdAuthenticate,
+		Args: CmdArgsAuthenticate{AccessToken: accessToken},
+	}); err != nil {
+		return CmdRsAuthenticate{}, err
+	} else {
+		return res.(CmdRsAuthenticate), nil
+	}
 }
 
-func (c *clientImpl) Logger() log.Logger {
-	return c.logger
+func (c *Client) GetGuild(guildID snowflake.ID, timeout int) (CmdRsGetGuild, error) {
+	args := CmdArgsGetGuild{
+		GuildID: guildID,
+	}
+
+	if timeout != 0 {
+		args.Timeout = timeout
+	}
+
+	if res, err := c.send(Message{
+		Cmd:  CmdGetGuild,
+		Args: args,
+	}); err != nil {
+		return CmdRsGetGuild{}, err
+	} else {
+		return res.(CmdRsGetGuild), nil
+	}
 }
 
-func (c *clientImpl) ServerConfig() ServerConfig {
-	return c.serverConfig
+func (c *Client) GetGuilds() ([]PartialGuild, error) {
+	if res, err := c.send(Message{
+		Cmd:  CmdGetGuilds,
+		Args: struct{ CmdArgs }{},
+	}); err != nil {
+		return nil, err
+	} else {
+		return res.(CmdRsGetGuilds).Guilds, nil
+	}
 }
 
-func (c *clientImpl) User() discord.User {
-	return c.user
+func (c *Client) GetChannel(channelID snowflake.ID) (CmdRsGetChannel, error) {
```

Comment:
All these methods should return `(*T, error)`

Reply:
Did you want this on Authorize as well?
eg. from `Authorize(scopes []discord.OAuth2Scope, rpcToken string, username string) (string, error)`
to `Authorize(scopes []discord.OAuth2Scope, rpcToken string, username string) (*string, error)`
or `Authorize(scopes []discord.OAuth2Scope, rpcToken string, username string) (*CmdRsAuthorize, error)`

Reply:
yes

Reply:
Resolved in [4ef8ffc](https://github.com/dennisrijsdijk/disgo/commit/4ef8ffc03ba58f70cb22951e155ddf66e18ef4e6)